### ISSUE TITLE
Calculate "distance" of user from your network

### ISSUE THREAD

**Issue by trwnh** [tootsuite/mastodon#11163]
### Pitch

Recursively crawl `following` collection of your own `following` collection, if published, up to an arbitrary depth (n=3? n=6?). Construct a graph from the discovered relationships. Calculate the shortest distance to each user discovered; let *d* be this shortest distance (e.g. with Dijkstra's Algorithm). Display *d* on profiles, if available.

### Motivation

It would offer users more information about how close they are to a certain person, e.g. "friend of a friend" semantics. It could also be used in the future with object capabilities to establish a proper web of trust where users within a certain distance are allowed to communicate with you, while distant users ("randos") are greylisted and must have their follows/mentions approved.

### Notes

- It should be noted that `following` is perhaps not 100% equivalent to trust, but it is currently the only metric available for establishing distance. It may be worth considering using a different collection for this purpose, e.g. a `trusts` collection containing IDs or public keys that are signed by an actor's own key.
- It should also be noted that, since the `following` collection is not always available (e.g. in implementations that do not publish this collection, or for Mastodon users who have enabled "hide your network"), there will always be some users who may in actuality be within a certain distance from you, but this information cannot be derived. That is probably acceptable in the short term, but it does provide more support for having separate collection(s) in the long term with more explicit purpose, e.g. `trusts`, `knows`, or `friends` that are published intentionally.

**Comment by Cassolotl** [comment]
"Follow" and "trust" aren't really equivalent, since sometimes people follow people they want to keep an eye on in a "know your enemy" kind of way.

**Comment by trwnh** [comment]
Yeah, it's important to keep that in mind (note 1). It's the main limitation of using the `following` collection, but it's also the only thing that exists right now, aside from `followers`. This is why it'd be better to have separate collections (the latter half of note 2).

**Comment by trwnh** [comment]
still worth exploring at a future point. i'd really like to know if someone is a totally random person or if we have a mutual link somewhere.
